test/unittests/eof_validation_test.cpp

```diff
@@ -1211,3 +1211,198 @@ TEST_F(eof_validation, max_nested_containers)
 
     add_test_case(code, EOFValidationError::success);
 }
+
+// Summary of validity of combinations of referencing instructions with instructions inside
+// referenced containers.
+// Rows are instructions referencing subcontainers or rules for top-level container.
+// Columns are instructions inside referenced subcontainer or none if it ends with REVERT/INVALID.
+//
+//    |                | STOP   | RETURN | RETURNCONTRACT | none |
+//    | -------------- | ------ | -------| -------------- | ---- |
+//    | top-level      | +      | +      | -              | +    |
+//    | EOFCREATE      | -      | -      | +              | +    |
+//    | RETURNCONTRACT | +      | +      | -              | +    |
+//    | both           | -      | -      | -              | +    |
+
+TEST_F(eof_validation, toplevel_container_stop)
+{
+    const auto code = bytecode{OP_STOP};
+    const auto container = eof_bytecode(code, 0);
+
+    add_test_case(container, EOFValidationError::success);
+}
+
+TEST_F(eof_validation, toplevel_container_return)
+{
+    const auto code = ret(0, 0);
+    const bytecode container = eof_bytecode(code, 2);
+
+    add_test_case(container, EOFValidationError::success);
+}
+
+TEST_F(eof_validation, toplevel_container_returncontract)
+{
+    const auto code = returncontract(0, 0, 0);
+    const auto container = eof_bytecode(code, 2).container(eof_bytecode(OP_INVALID));
+
+    add_test_case(container, EOFValidationError::success);
```

Comment:
Shouldn't this case fail validation?

Reply:
This looks to be not in the current spec https://github.com/ipsilon/eof/pull/86/files and probably it should be.

On the other hand there's no notion of "top-level" container, it's kinda something we have only in the tests and it's not clearly defined. Like, creation transaction is top-level container, but deployed contract is also top-level container...

Reply:
My hunch is that validation of top-level container, i.e. "general definition of validation" should allow either RETURNCONTRACT or STOP/RETURN. For creation transaction it's an extra rule that its toplevel container is initcontainer. For deployed containers (ones referenced by RETURNCONTRACT), it's a rule that they are runtime containers.

It's already specified like this I believe.

Reply:
New idea is to be explicit in validation tests whether we validate runtime container or initcontainer (extra flag for each case), then there will be no confusing "top-level container" tests.

Reply:
This may require another refactoring and I think we can get rid of `validate_eof_get_container_kind()` (we'll move toplevel container kind checks into `validate_eof()` and creation tx / eofcreate / txcreate will request to validate it as initcontainer.

Reply:
Makes sense :+1:

Reply:
Refactoring done, this case now succeeds in initcode mode.